File: UtilityBasedAICPP/ConsoleApplication1/CoverUtilityCalculator.cpp
#include "CoverUtilityCalculator.h"

namespace UtilityBasedAI
{
    float CoverUtilityCalculator::CalculateUtility(AIController* character)
    {
        switch (character->CurrentEnemyState)
        {
        case EnemyState::Idle:
            return (100 - character->Health) * 0.1f;
        case EnemyState::Aggressive:
            return character->DistanceToEnemy + (100 - character->Health);
        case EnemyState::CallingForBackup:
            return (100 - character->DistanceToEnemy) + (100 - character->Health) * 1.5f;
        case EnemyState::Retreating:
            return (100 - character->DistanceToEnemy) + (100 - character->Health) * 0.5f;
        default:
            return 0;
        }
    }
}

File: UtilityBasedAI/UtilityBasedAI/CoverUtilityCalculator.cs
namespace UtilityBasedAI
{
    public class CoverUtilityCalculator : IUtilityCalculator
    {
        public float CalculateUtility(CharacterAI character)
        {
            switch (character.EnemyState)
            {
                case EnemyState.Idle:
                    return (100 - character.Health) * 0.1f;
                case EnemyState.Aggressive:
                    return character.DistanceToEnemy + (100 - character.Health);
                case EnemyState.CallingForBackup:
                    return character.DistanceToEnemy + (100 - character.Health) * 1.5f;
                case EnemyState.Retreating:
                    return character.DistanceToEnemy + (100 - character.Health) * 0.5f;
                default:
                    return 0;
            }
        }
    }
}